node/src/components/block_proposer.rs

```diff
@@ -97,49 +98,120 @@ impl<REv> ReactorEventT for REv where
 {
 }
 
-/// Deploy buffer.
-#[derive(DataSize, Debug, Clone)]
-pub(crate) struct DeployBuffer {
+/// Stores the internal state of the BlockProposer.
+#[derive(DataSize, Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockProposerState {
     pending: DeployCollection,
     proposed: ProtoBlockCollection,
     finalized: ProtoBlockCollection,
+}
+
+impl Display for BlockProposerState {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "(pending:{}, proposed:{}, finalized:{})",
+            self.pending.len(),
+            self.proposed.len(),
+            self.finalized.len()
+        )
+    }
+}
+
+impl BlockProposerState {
+    // When deserialized, we will update the instance with finalized blocks from the linear chain
+    pub(crate) fn update_finalized_blocks(
+        &mut self,
+        finalized: ProtoBlockCollection,
+        current_instant: Timestamp,
+    ) {
+        self.prune(current_instant);
+        let mut finalized = finalized;
+        let _ = prune::prune_blocks(&mut finalized, current_instant);
+        self.finalized.extend(finalized);
+    }
+    /// Prunes expired deploy information from the BlockProposerState, returns the total deploys
+    /// pruned
+    pub(crate) fn prune(&mut self, current_instant: Timestamp) -> usize {
+        let pending = prune::prune_deploys(&mut self.pending, current_instant);
+        let proposed = prune::prune_blocks(&mut self.proposed, current_instant);
+        let finalized = prune::prune_blocks(&mut self.finalized, current_instant);
+        pending + proposed + finalized
+    }
+}
+
+mod prune {
+    use super::*;
+    /// Prunes expired deploy information from an individual DeployCollection, returns the total
+    /// deploys pruned
+    pub(super) fn prune_deploys(
+        deploys: &mut DeployCollection,
+        current_instant: Timestamp,
+    ) -> usize {
+        let initial_len = deploys.len();
+        deploys.retain(|_hash, header| !header.expired(current_instant));
+        initial_len - deploys.len()
+    }
+    /// Prunes expired deploy information from each ProtoBlockCollection, returns the total
+    /// deploys pruned
+    pub(super) fn prune_blocks(
+        blocks: &mut ProtoBlockCollection,
+        current_instant: Timestamp,
+    ) -> usize {
+        let mut pruned = 0;
+        let mut remove = Vec::new();
+        for (block_hash, deploys) in blocks.iter_mut() {
+            pruned += prune_deploys(deploys, current_instant);
+            if deploys.is_empty() {
+                remove.push(*block_hash);
+            }
+        }
+        blocks.retain(|k, _v| !remove.contains(&k));
```

Comment:
Since `blocks` is a map and `remove` is a `Vec`, would it be faster to iterate over the latter instead?
```rust
remove.iter().for_each(|k| blocks.remove(k));
```

Reply:
Got it, thanks.